### ISSUE TITLE
attempting to GC indexes: clearing index 2: command is too large

### ISSUE THREAD

**Issue by dankinder** [cockroachdb/cockroach#61206]
**Describe the problem**

On truncating a table with about 180GB of data, the GC got this error:
`attempting to GC indexes: clearing index 2: command is too large: 120227141 bytes (max: 67108864)`

This was data we had imported (via IMPORT INTO ... CSV) within the past day.

Note, this is using v21.1.0-alpha3 in order to get [this fix](https://github.com/cockroachdb/cockroach/pull/59204), or else our S3 reads time out.

I have a debug zip exported if you want me to upload it in the support portal.

**Environment:**
 - CockroachDB version v21.1.0-alpha3
 - Server OS: Centos 6

**Comment by ajwerner** [comment]
Is this using interleaving somehow?

**Comment by ajwerner** [comment]
There is a cluster setting with the max command size; that may unstick you but it carries modest risk. Definitely set it back down afterwards. Fortunately this isn't attempting to do anything totally nuts. The other option is to decrease the max range size for the database to, say, 32 MiB (or just the default one if that database is gone now). That's a safer choice but will incur more load on the cluster as splitting and merging happens.

**Comment by dankinder** [comment]
Interestingly it seems like that data still got dumped regardless... so I guess there's nothing about our cluster that needs remediation per se. But I assume you would want to prevent this from happening in the future. Nothing about the dataset was really unusual.

And no, no interleaving. This was a really simple dataset with one table, a few columns with one INT primary key.

**Comment by ajwerner** [comment]
"Still got dumped" as in gc happened?

Also, that error, where are you seeing it? Is it in the logs or did it return from truncate itself?

**Comment by dankinder** [comment]
What I mean is, the number of live bytes dropped dramatically, so seemingly most of the data got cleared if not all of it.

This error is not on the TRUNCATE job, it's on the GC job that followed it, i.e. it's `GC for TRUNCATE TABLE <the table>` that failed.

If that failure does leave data leftover, will it eventually get cleared in the normal compaction process?

**Comment by ajwerner** [comment]
It's bad that that job failed. We've had recent discussions about whether we should ever let that job fail. 

https://github.com/cockroachdb/cockroach/issues/55740
https://github.com/cockroachdb/cockroach/issues/59542
https://github.com/cockroachdb/cockroach/issues/59788#issuecomment-773662596

I can help you to re-start that job. In the meantime, can you grab a copy of its record before it gets deleted by the system? That'd be:

```sql
SELECT id, status, created, encode(payload, 'hex'), encode(progress, 'hex') FROM system.jobs WHERE id = <relevant job id>;
```

**Comment by ajwerner** [comment]
What I don't understand is why the GC job would be sending a large raft command. The clear range operation it uses should be small.

**Comment by ajwerner** [comment]
Are your keys somehow absolutely gigantic?

**Comment by ajwerner** [comment]
I think I've got a lead on this one. We'll need to do some manual things to recover the job. Thanks for the bug report!

**Comment by ajwerner** [comment]
Actually we're still pretty confused. Do you have any more intel on the structure of these tables to share?

**Comment by dankinder** [comment]
Yeah so I can at least say it's an extremely simple table, basically a few INT columns that together are the primary key. No other columns, no indexes.

I just sent a debug zip through the support portal and tagged this issue.

**Comment by dankinder** [comment]
If y'all couldn't find anything and don't intend to investigate further (because it's an alpha), it's okay if you want to close this, we're good as far as our cluster.

**Comment by erikgrinaker** [comment]
I think we've found the cause of this, submitted a fix in #74674.
